src/main.rs

```diff
@@ -34,321 +33,323 @@ extern crate xi_core_lib;
-                    edit_view.set_view_id(value.as_str().unwrap());
-                });
+                let view_id = value.clone().as_str().unwrap().to_string();
+                let mut state = state.lock().unwrap();
+                let handle = handle.clone();
+                state.focused = view_id.clone();
+                state.views.insert(view_id.clone(),
+                    ViewState {
+                        id: 0,
+                        filename: filename.clone(),
+                        handle: handle.clone(),
+                    }
+                );
+                UiMain::send_ext(&handle.lock().unwrap(), edit_view, EditViewCommands::Core(core));
+                UiMain::send_ext(&handle.lock().unwrap(), edit_view, EditViewCommands::ViewId(view_id));
             }
         );
     }
 
     fn handle_cmd(&self, method: &str, params: &Value) {
-        let mut state = self.state.borrow_mut();
         match method {
-            "update" => state.edit_view.apply_update(&params["update"]),
-            "scroll_to" => state.edit_view.scroll_to(params["line"].as_u64().unwrap() as usize),
+            "update" => self.send_view_cmd(EditViewCommands::ApplyUpdate(params["update"].clone())),
+            "scroll_to" => self.send_view_cmd(EditViewCommands::ScrollTo(params["line"].as_u64().unwrap() as usize)),
             "available_themes" => (), // TODO
             "available_plugins" => (), // TODO
+            "available_languages" => (), // TODO
             "config_changed" => (), // TODO
+            "language_changed" => (), // TODO
             _ => println!("unhandled core->fe method {}", method),
         }
-        // TODO: edit view should probably handle this logic
-        self.invalidate();
-    }
-
-    pub fn invalidate(&self) {
-        self.handle.borrow().invalidate();
     }
 }
 
-fn create_main(core: Core) -> Result<WindowHandle, Error> {
-    let main_state = MainWinState::new();
-    let main_win = Rc::new(MainWin::new(core, main_state));
-    let main_win_handler = MainWinHandler {
-        win: main_win,
-    };
-
-    let menubar = menus::create_menus();
-
-    let mut builder = WindowBuilder::new();
-    builder.set_handler(Box::new(main_win_handler));
-    builder.set_title("xi-editor");
-    builder.set_cursor(Cursor::IBeam);
-    builder.set_menu(menubar);
-    let window = builder.build().unwrap();
-    Ok(window)
-}
-
 #[derive(Clone)]
-struct MyHandler {
-    runloop: RunLoopHandle,
-    win_handle: Arc<Mutex<Option<IdleHandle>>>,
+struct AppDispatcher {
+    app: Arc<Mutex<Option<App>>>,
 }
 
-impl MyHandler {
-    fn new(runloop: RunLoopHandle) -> MyHandler {
-        MyHandler {
-            runloop,
-            win_handle: Default::default(),
+impl AppDispatcher {
+    fn new() -> AppDispatcher {
+        AppDispatcher {
+            app: Default::default(),
         }
     }
+
+    fn set_app(&mut self, app: &App) {
```

Comment:
If you do keep the `Arc<Mutex<>>`, then this doesn't need to be `&mut`. Or it's possible you get rid of the former and keep the latter.